### DISCUSSION TITLE
possible to toggle showing implicits

### DISCUSSION THREAD

**Discussion by stumash** [scalameta/nvim-metals#188] [Q&A] [answered]
I see that there are configuration options on setup for showing `implicit` arguments and conversions.

Is there a way to toggle these options _while `metals` is running_?

**Comment by ckipp01** [comment, marked as answer]
So in theory, yes there is. These settings are simply passed to metals as `UserConfiguration` settings that Metals can be notified of using a `didChangeConfiguration` notification. The tricky part is that this currently is really only sent once, right after the initialize phase. Editors that have a configuration file of sorts that can be updated on the fly by the editor can change the values and re-send the notification, making this easy to do. However, since the configuration for nvim-metals is read in once and then that's sort of it, we'll have to play around a bit with how to achieve this. It's doable, but the persisting may be the problem. Since if you close it with the setting on, but don't have it set to on in your configuration, you'll start with it off again. I'll create a feature request out of this since it'd be a great feature to have.

**Reply by stumash** [reply]
hey @ckipp01 thanks for the reply!

FWIW, I don't think that persisting the setting once toggled would be desired. Like many toggle-able settings in vim, whenever you restart you expect the startup-settings to be in place. So you wouldn't have to handle that IMO, you'd just have to make the `didChangeConfiguration` callable in some way or another.

**Reply by ckipp01** [reply]
> FWIW, I don't think that persisting the setting once toggled would be desired. Like many toggle-able settings in vim, whenever you restart you expect the startup-settings to be in place. So you wouldn't have to handle that IMO, you'd just have to make the didChangeConfiguration callable in some way or another.

yea that make sense. This should be doable 👍🏼

**Reply by stumash** [reply]
Awesome. I'm happy to contribute if you could point me at some code 😬 ?

Also.... total side note:

1. Do you know if the `implicit` info in metals being turned on slows it down?
2. Any idea what I need to do to make this [this PR](https://github.com/tree-sitter/tree-sitter-scala/pull/29) get merged? It adds super-simple keyword-only highlighting to `scala-tree-sitter`.

thanks for all your work in the scala + vim community

**Reply by ckipp01** [reply]
> Awesome. I'm happy to contribute if you could point me at some code 😬 ?

So the actual notification that would need to be sent to Metals can be seen here: https://github.com/scalameta/nvim-metals/blob/32fc0b137407b169e01cbc762844cdbfacacfe4f/lua/metals/setup.lua#L315

So the general flow would probably be a function that could grab the current settings, change the value of the specific one that you want to toggle, and then simply send in the notification. I'd have to play around a bit more with it to give more details.

> Do you know if the implicit info in metals being turned on slows it down?

This won't really slow Metals down for you. Especially since this info is only going to be shown in the hover anyways. Some other settings, like showing the super method code lenses can for sure slow Metals down though.

> Any idea what I need to do to make this this PR get merged? It adds super-simple keyword-only highlighting to scala-tree-sitter.

No idea. If it's been a while, maybe kindly ping Max to review it as I believe he's the only one with merge power.
